File: java/313.java
class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        Arrays.sort(primes);
        int[] ans = new int[n];
        ans[0] = 1;
        int[] id = new int[primes.length];
        int[] fac = primes.clone();
        int v = 1;
        while(v < n){
            int min = min(fac);
            ans[v++] = min;
            for(int i = 0;i <primes.length;i++){
                if(min == fac[i]){
                    fac[i] = ans[id[i]+1] * primes[i];
                    id[i] ++;
                }
            }
        }
        return ans[n-1];
    }
    public int min(int[] nums){
        int min = Integer.MAX_VALUE;
        for(int num:nums){
            min = Math.min(num,min);
        }
        return min;
    }
}

File: go/313.go
func nthSuperUglyNumber(n int, primes []int) int {
	dp := make([]int, n)
	dp[0] = 1

	// sort.Ints(primes)
	l := len(primes)
	id := make([]int, l)
	fac := make([]int, l)
	for i, v := range primes {
		fac[i] = v
	}
	for i := 1; i < n; i++ {
		min := min(fac)
		dp[i] = min
		for j, v := range fac {
			if min == v {
				fac[j] = dp[id[j]+1] * primes[j]
				id[j] = id[j] + 1
			}
		}
	}

	return dp[n-1]
}

func min(nums []int) int {
	x := nums[0]
	for _, v := range nums {
		if v < x {
			x = v
		}
	}
	return x
}